src/codelabs/null-safety.md

```diff
@@ -0,0 +1,357 @@
+---
+title: "Null safety"
+description: Learn about and practice writing null safe code in DartPad!
+js: [{url: 'https://dartpad.dev/inject_embed.dart.js', defer: true}]
+---
+{% assign useIframe = false -%}
+<?code-excerpt replace="/ *\/\/\s+ignore_for_file:[^\n]+\n//g; /(^|\n) *\/\/\s+ignore:[^\n]+\n/$1/g; /(\n[^\n]+) *\/\/\s+ignore:[^\n]+\n/$1\n/g"?>
+<?code-excerpt plaster="none"?>
+<style>
+{% comment %}
+TODO(chalin): move this into one of our SCSS files
+{% endcomment -%}
+iframe[src^="https://dartpad"] {
+  border: 1px solid #ccc;
+  margin-bottom: 1rem;
+  min-height: 150px;
+  resize: vertical;
+  width: 100%;
+}
+</style>
+
+This codelab teaches you about Dart’s null safe type system, which was introduced in Dart SDK 2.12. When you opt into null safety, types in your code are non-nullable by default, meaning that values can’t be null unless you say they can be.
+
+Using embedded DartPad editors, you can test your knowledge by running example code and completing exercises. To get the most out of this codelab, you should have some knowledge of [basic Dart syntax](/samples).
+
+This codelab covers the following material:
+
+* Nullable and non-nullable types.
+* How and when to use the null-aware operators: ? and !
+* How the late keyword affects variables and initialization.
+
+{{site.alert.note}}
+  This page uses embedded DartPads to display examples and exercises.
+  {% include dartpads-embedded-troubleshooting.md %}
+{{site.alert.end}}
+
+
+## What is null safety?
+
+Dart's null safe type system makes types in your code non-nullable by default,
+meaning that values can't be null unless you say they can be. Null safety helps
+prevent null reference errors. 
+
+Here are some core principles
+of sound null safety in the Dart language:
+
+-   All types are non-nullable by default.
+-   Values can't be null unless you say they can be.
+-   Finding null errors at edit time is much better than finding them at run
+    time.
+
+## Non-nullable types
+
+All types are non-nullable by default. For example, if you have a variable of
+type String, it will always contain a string.
+
+## Nullable types
+
+If you want a variable of type String to accept any string, or the value null
```

Comment:
```suggestion
If you want a variable of type `String` to accept any string, or the value null
```

I'll stop suggesting, but, in general, I tend to add backticks around every type and piece of code in any given paragraph.